More Structured Job I/O Tracking

jmchilton opened the pull request:
- Track messages generated tool stdio tags in a JSON blob and don't prepend them to the tool's reported stdout/stderr. Report them separately in the UI.
- Track job script stdout and stderr separately from the tool's stdout and stderr in the database. Should make our ability to report information about dependency resolution and metadata generation and lot better.

Builds on #6685 #6338. Pulled out the tracking branch #7058 for issue #7050.

bernt-matthias committed:
- check regexp also if FATAL was detected by exit_code
- introduced functions for checking regex against stream
- check regex line by line instead of on the complete stream
  (the original behavior is coded in check_output_regex which
  could be removed if linewise is OK)

bernt-matthias committed:
also check all regexes

bernt-matthias committed:
added test and changed modification of stdout/err

before stdout and stderr were modified while testing for exit codes and
regexes. this had the potential problem that later tests also checked
this output which might lead to confusions.

furthermore a marker line has been introduces which differentiates
Galaxy and original tool output

bernt-matthias committed:
fixed linting and copy paste errors

jmchilton committed:
Don't determine job failure from stderr, that makes no sense anymore.

jmchilton committed:
More consistent exit_code handling across runners.

jmchilton committed:
More consistent job finish/resubmit handling across local and async runners.

jmchilton committed:
De-duplicate constants and parameters for shrink_stream in jobs code.

jmchilton committed:
Move model/Tool interactions out of galaxy.jobs.output_checker.

This means output_checker can be placed into galaxy-lib without dependency problems, and we can evaluate if a tool passed without access to a job object or a Tool object as long as we have a description of the job outputs and stdio objects (already in galaxy-lib).

This is a prelude to being able to do job output checking in Pulsar without Galaxy internals.

jmchilton commented:
Delaying this to 19.05 but it is ready to go, if someone wants to be nice and merge it before the release is branched just move the tag back.

mvdbeek commented on lib/galaxy/tools/verify/interactor.py:
```diff
@@ -817,11 +817,36 @@ def register_exception(e):
         "stdout": "Standard output of the job",
         "stderr": "Standard error of the job",
     }
+    # TODO: Only hack the stdio like this for older profkle, for newer tool profiles
```
s/profkle/profile

jmchilton replied:
will rebase with a fix, thanks

mvdbeek commented on lib/galaxy/jobs/output_checker.py:
```diff
@@ -14,11 +14,45 @@
+    return max_error_level
+
+
+def check_output_regex_byline(job_id_tag, regex, stream, stream_append, max_error_level):
```
This  isn't used AFAICT, drop it ?

jmchilton replied:
will rebase with a fix, thanks

jmchilton commented:
I don't think this quite works at all like I expected... putting it back into WIP. Might be good to have tests that somehow proved the streams were being separately properly.

mvdbeek commented on lib/galaxy/jobs/__init__.py:
```diff
@@ -993,7 +993,7 @@ def _fix_output_permissions(self):
                 util.umask_fix_perms(path, self.app.config.umask, 0o666, self.app.config.gid)
 
-    def fail(self, message, exception=False, stdout="", stderr="", exit_code=None):
+    def fail(self, message, exception=False, tool_stdout="", tool_stderr="", exit_code=None, job_stdout=None, job_stderr=None):
```
Need to update `job_wrapper.fail(stdout=...)` kubernetes.py and pulsar.py for this change as well.

jmchilton replied:
indeed - how did I miss that. will rebase with a fix, thanks

mvdbeek commented on lib/galaxy/jobs/output_checker.py:
```diff
@@ -14,11 +14,45 @@
+    return max_error_level
+
+
+def check_output(stdio_regexes, stdio_exit_codes, stdout, stderr, tool_exit_code, job_id_tag):
```
`state = DETECTED_JOB_STATE.OK` on line 69 has no effect, we could drop it.

jmchilton replied:
I like removing the ``else``s inside the main ``if`` more? It is more clear this way what the default value is and that it always has a default value. I'm going to do that instead - hope that is okay.

mvdbeek commented on lib/galaxy/jobs/output_checker.py:
```diff
@@ -149,32 +180,35 @@ def check_output(tool, stdout, stderr, tool_exit_code, job):
-        job.set_streams(stdout, stderr)
-
-    return state
+    return state, stdout, stderr, job_messages
```
Can we bump up the log.warning above to log.exception ?
This seems like something serious that we should be fixed if it ever really happens (plus we can drop the `traceback.format_exc()` part).

jmchilton replied:
will do in a rebase when fixing up default state, thanks!

mvdbeek commented on lib/galaxy/jobs/output_checker.py:
```diff
@@ -14,11 +14,45 @@
+    for line in reversed(stream.split("\n")):
+        regex_match = re.search(regex.match, line, re.IGNORECASE)
+        if regex_match:
+            rexmsg = __regex_err_msg(regex_match, regex)
```
we don't use it, but it should be `__regex_err_msg(regex_match, stream_name, regex)` if we don't drop it

mvdbeek commented on lib/galaxy/jobs/output_checker.py:
```diff
@@ -149,32 +180,35 @@ def check_output(tool, stdout, stderr, tool_exit_code, job):
     # Otherwise, we'll take the first 256 characters of the match.
     if None is not regex.desc:
-        err_msg += regex.desc
+        desc += regex.desc
```
I guess `if regex.desc is not None` would read a bit better ?

jmchilton replied:
Will fix in a rebase, thanks.

mvdbeek commented on lib/galaxy/jobs/runners/__init__.py:
```diff
@@ -437,22 +437,62 @@ def mark_as_resubmitted(self, job_state, info=None):
+            job = job_state.job_wrapper.get_job()
+            exit_code = job_state.read_exit_code()
+
+            tool_stdout = tool_stderr = None
```
This can't be reached, and `None` would blow up in `check_output_regex`.

jmchilton replied:
... will fix in a rebase - thanks

mvdbeek commented on lib/galaxy/model/__init__.py:
```diff
@@ -252,6 +263,27 @@ def log_str(self):
+        return stderr
+
+    def set_stderr(self, stderr):
+        raise NotImplementedError("Attempt to set stdout, must set tool_stderr or job_stderr")
```
We need to change https://github.com/galaxyproject/galaxy/blob/355b48c361e68a7d9fbe9fe86de7aa2313e5d608/lib/galaxy/webapps/galaxy/controllers/admin.py#L1647 and https://github.com/galaxyproject/galaxy/blob/355b48c361e68a7d9fbe9fe86de7aa2313e5d608/lib/galaxy/jobs/handler.py#L302 too, right ?

mvdbeek reviewed (commented):
This is awesome, but I think there's some minor things that need fixing.

jmchilton committed:
Track job failure reasons in a more structured way.

jmchilton committed:
Remove unused code spotted by mvdbeek.

jmchilton committed:
Simplify, clear up output_checker code a bit.

Based on suggestions from mvdbeek.

mvdbeek commented:
> Might be good to have tests that somehow proved the streams were being separately properly.

How about a integration test that uses a custom DEFAULT_JOB_FILE_TEMPLATE to inject some messages on stdout/stderr ?

jmchilton committed:
Seperate tool and job script standard I/O.

This should allow Galaxy to reason about what the tool reports vs. what dependency resolution and metadata reports, allow us to enable logging during metadata collection, and allow us to reason about the tool's I/O from inside the job script during cleanup/metadata collection at the end.

jmchilton commented:
@mvdbeek Thanks for the detailed review, I'll admit without that test this was pretty half baked. I've added a test now and I think it is working?

mvdbeek commented:
Awesome, yeah, looks great. I have a feeling that we might also be able to get rid of `capture_stdout_stderr` introduced for kubernetes.

mvdbeek commented:
Thanks again @bernt-matthias @jmchilton, that was much needed!
